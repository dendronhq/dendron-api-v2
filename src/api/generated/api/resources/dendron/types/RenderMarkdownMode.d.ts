/**
 * This file was auto-generated by Fern from our API Definition.
 */
/**
 * The rendering mode.
 */
export declare type RenderMarkdownMode = "markdown" | "gfm";
export declare const RenderMarkdownMode: {
    readonly Markdown: "markdown";
    readonly Gfm: "gfm";
};
