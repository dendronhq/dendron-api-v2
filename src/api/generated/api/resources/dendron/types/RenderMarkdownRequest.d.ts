/**
 * This file was auto-generated by Fern from our API Definition.
 */
import { FernApi } from "../../../..";
export interface RenderMarkdownRequest {
    /** The Markdown text to render in HTML. */
    text: string;
    /** The rendering mode. */
    mode?: FernApi.RenderMarkdownMode;
    /** The repository context to use when creating references in `gfm` mode.  For example, setting `context` to `octo-org/octo-repo` will change the text `#42` into an HTML link to issue 42 in the `octo-org/octo-repo` repository. */
    context?: string;
}
