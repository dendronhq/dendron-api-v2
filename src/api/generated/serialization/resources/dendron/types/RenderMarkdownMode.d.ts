/**
 * This file was auto-generated by Fern from our API Definition.
 */
import * as serializers from "../../..";
import { FernApi } from "../../../..";
import * as core from "../../../../core";
export declare const RenderMarkdownMode: core.serialization.Schema<serializers.RenderMarkdownMode.Raw, FernApi.RenderMarkdownMode>;
export declare namespace RenderMarkdownMode {
    type Raw = "markdown" | "gfm";
}
