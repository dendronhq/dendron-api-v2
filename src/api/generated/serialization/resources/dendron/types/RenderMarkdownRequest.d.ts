/**
 * This file was auto-generated by Fern from our API Definition.
 */
import * as serializers from "../../..";
import { FernApi } from "../../../..";
import * as core from "../../../../core";
export declare const RenderMarkdownRequest: core.serialization.ObjectSchema<serializers.RenderMarkdownRequest.Raw, FernApi.RenderMarkdownRequest>;
export declare namespace RenderMarkdownRequest {
    interface Raw {
        text: string;
        mode?: serializers.RenderMarkdownMode.Raw | null;
        context?: string | null;
    }
}
