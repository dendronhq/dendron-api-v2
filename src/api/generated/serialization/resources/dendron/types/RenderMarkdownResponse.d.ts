/**
 * This file was auto-generated by Fern from our API Definition.
 */
import * as serializers from "../../..";
import { FernApi } from "../../../..";
import * as core from "../../../../core";
export declare const RenderMarkdownResponse: core.serialization.ObjectSchema<serializers.RenderMarkdownResponse.Raw, FernApi.RenderMarkdownResponse>;
export declare namespace RenderMarkdownResponse {
    interface Raw {
        content: string;
    }
}
