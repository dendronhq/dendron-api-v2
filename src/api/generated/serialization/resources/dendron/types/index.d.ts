export * from "./RenderMarkdownRequest";
export * from "./RenderMarkdownResponse";
export * from "./RenderMarkdownMode";
