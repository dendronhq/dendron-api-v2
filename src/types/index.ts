
import { GrayMatterFile } from 'gray-matter';

export type MatterFile = {
  fname: string
  fpath: string
} & GrayMatterFile<string>;

export type FileData = Pick<GrayMatterFile<string>, "content" | "data">

export type NoteData = {
  id: string;
  title: string;
  tags: string[];
  updated: number;
  created: number;
}