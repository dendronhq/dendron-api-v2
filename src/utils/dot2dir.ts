import fs from "fs-extra";
import matter from "gray-matter";
import _ from "lodash";
import path from "path";
import { FileData } from "../types";
import { readDirRecursive } from "./file";

export type FileType = "file" | "index";

export function fnames2FileTypeMap(filenames: string[]): Map<string, FileType> {
  const fileTypes = new Map<string, FileType>();

  // Populate fileTypes with all filenames as keys and "file" as default value
  for (const filename of filenames) {
    fileTypes.set(filename, "file");
  }

  for (const filename of filenames) {
    const parts = filename.split(".");

    // Iterate through prefixes and update fileTypes if a parent is found
    for (let i = parts.length - 1; i > 0; i--) {
      const prefix = parts.slice(0, i).join(".");

      if (fileTypes.has(prefix)) {
        fileTypes.set(prefix, "index");
        break;
      }
    }
  }

  return fileTypes;
}

export function fname2FilePath(fname: string, fileType: FileType) {
  const _path = fname.replace(/\./g, '/');
  if (fileType === "file") {
    return _path + ".md"
  } else {
    return _path + "/index.md"
  }
}

/**
 *  Materialize files and folders in the file system
 */
export async function materializeFnames2FilesAndFolders(fileMappings: Map<string, FileType>,
  opts: { fnameDataMapping: Map<string, FileData>, baseDir: string, transformTags?: boolean }) {

  // make sure all folders exist
  const fileMappingArray = Array.from(fileMappings.entries());
  await Promise.all(_.map(fileMappingArray, ([fname, fileType]) => {
    const _fpath = path.join(opts.baseDir, fname2FilePath(fname, fileType))
    return fs.ensureDir(path.dirname(_fpath))
  }));

  // create all files
  await Promise.all(_.map(fileMappingArray, ([fname, fileType]) => {
    const _fpath = path.join(opts.baseDir, fname2FilePath(fname, fileType))

    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion, prefer-const
    let { content, data } = opts.fnameDataMapping.get(fname)!

    // regex that matches all tags with dots and replaces them with forward slash
    if (opts.transformTags) {
      ({ content, data } = transfromTags({ content, data }))
    }

    const matterResult = matter.stringify(content, data);
    return fs.writeFile(_fpath, matterResult)
  }));
}

export function readFilesRecursively(dir: string,
  filterFns: ((filename: string) => boolean)[] = []): string[] {
  return readDirRecursive(dir, '', filterFns)
}

function transfromTags(opts: { data: any, content: string }) {
  const { data, content } = opts;
  let cleanTags = _.get(data, 'tags', []);
  // TODO: we have bad casees like 
  // tags:
  //   - - - tags.dendron.lvl2
  if (cleanTags && _.isArray(cleanTags) && _.isString(cleanTags[0])) {
    cleanTags = cleanTags.map((tag: string) => {
      return tag.replace(/\./g, '/')
    });
  }
  data.tags = cleanTags;
  const cleanContent = content.replace(/#([^\s]+)/g, (_, tag) => {
    return `#${tag.replace(/\./g, '/')}`
  });
  return { data, content: cleanContent }
}

export function file2note(fpath: string) {
  const contents = fs.readFileSync(fpath, "utf8");
  const fname = path.basename(fpath).replace(/\.md$/, "")
  return { ...matter(contents), fname, fpath };
}