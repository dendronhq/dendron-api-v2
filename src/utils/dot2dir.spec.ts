import fs from "fs-extra";
import path from "path";
import { FileData } from '../types';
import { FileType, fnames2FileTypeMap, materializeFnames2FilesAndFolders } from './dot2dir';

/**
 * This is used for testing. Read a directory recursively 
 * @param dirPath 
 * @returns A map of file names and their contents
 */
function readDirectoryRecursive(dirPath: string): Map<string, string> {
  const filesMap = new Map<string, string>();
  const files = fs.readdirSync(dirPath);
  for (const file of files) {
    const filePath = path.join(dirPath, file);
    const stats = fs.statSync(filePath);

    if (stats.isDirectory()) {
      const subFilesMap = readDirectoryRecursive(filePath);
      subFilesMap.forEach((subFileContent, subFilePath) => {
        filesMap.set(subFilePath, subFileContent);
      });
    } else if (stats.isFile()) {
      const fileContent = fs.readFileSync(filePath, 'utf8');
      filesMap.set(filePath, fileContent);
    }
  }

  return filesMap;
}



describe('fnames2FileTypeMap', () => {
  test('should correctly identify file types', () => {
    const filenames = [
      'bar',
      'bar.two',
      'bar.alpha',
      'bar.alpha.one',
      'foo.two.three',
    ];
    const expected = new Map<string, FileType>([
      ['bar', 'index'],
      ['bar.two', 'file'],
      ['bar.alpha', 'index'],
      ['bar.alpha.one', 'file'],
      ['foo.two.three', 'file'],
    ]);

    expect(fnames2FileTypeMap(filenames)).toEqual(expected);
  });

  test('should handle empty input', () => {
    const filenames: string[] = [];
    const expected = new Map<string, FileType>();

    expect(fnames2FileTypeMap(filenames)).toEqual(expected);
  });

  test('should handle single-level filenames', () => {
    const filenames = ['foo', 'bar', 'baz'];
    const expected = new Map<string, FileType>([
      ['foo', 'file'],
      ['bar', 'file'],
      ['baz', 'file'],
    ]);

    expect(fnames2FileTypeMap(filenames)).toEqual(expected);
  });

  test('should handle multi-level filenames', () => {
    const filenames = ['a', 'a.b', 'a.b.c', 'a.b.c.d', 'a.b.d'];
    const expected = new Map<string, FileType>([
      ['a', 'index'],
      ['a.b', 'index'],
      ['a.b.c', 'index'],
      ['a.b.c.d', 'file'],
      ['a.b.d', 'file'],
    ]);

    expect(fnames2FileTypeMap(filenames)).toEqual(expected);
  });
});


describe('materializeFnames2FilesAndFolders', () => {
  const BASE_DIR = '/tmp/dendron';
  beforeEach(() => {
    fs.emptyDirSync(BASE_DIR);
  });

  test('should correctly materialize files and folders', async () => {
    const filenames = [
      'bar',
      'bar.two',
      'bar.alpha',
      'bar.alpha.one',
      'foo.two.three',
    ];
    const fnameDataMapping = {
      'bar': { content: 'bar', data: { id: 'bar-id' } },
      'bar.two': { content: 'bar.two', data: { id: 'bar-two-id' } },
      'bar.alpha': { content: 'bar.alpha', data: { id: 'bar-alpha-id' } },
      'bar.alpha.one': { content: 'bar.alpha.one', data: { id: 'bar-alpha-one-id' } },
      'foo.two.three': { content: 'foo.two.three', data: { id: 'foo-two-three-id' } },
    };
    const expected = [
      "/tmp/dendron/bar/alpha/index.md",
      "/tmp/dendron/bar/alpha/one.md",
      "/tmp/dendron/bar/index.md",
      "/tmp/dendron/bar/two.md",
      "/tmp/dendron/foo/two/three.md",
    ]

    await materializeFnames2FilesAndFolders(fnames2FileTypeMap(filenames), { baseDir: BASE_DIR, fnameDataMapping: new Map<string, FileData>(Object.entries(fnameDataMapping)) });
    const out = readDirectoryRecursive(BASE_DIR)
    expect(out).toMatchSnapshot();
    expect(Array.from(out.keys())).toEqual(expected)
  });

  describe('transformTags', () => {
    test('should replace dots in tags with slashes', async () => {
      const filenames = [
        'tags.fm-only',
        'tags.body-only',
        'tags.fm-and-body',
      ];
      const fnameDataMapping = {
        'tags.fm-only': { content: '', data: { tags: ['fm.one'] } },
        'tags.fm-mult': { content: '', data: { tags: ['fm.one', 'fm.two'] } },
        'tags.body-only': { content: '#body.one\n#body.one.two', data: {} },
        'tags.fm-and-body': { content: '#body.one', data: { tags: ['fm.one'] } },
      };
      await materializeFnames2FilesAndFolders(fnames2FileTypeMap(filenames),
        { baseDir: BASE_DIR, fnameDataMapping: new Map<string, FileData>(Object.entries(fnameDataMapping)), transformTags: true }
      );
      const out = readDirectoryRecursive(BASE_DIR)
      expect(out).toMatchSnapshot();
    });
  });

});



