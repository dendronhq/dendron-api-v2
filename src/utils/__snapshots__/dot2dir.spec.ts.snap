// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`materializeFnames2FilesAndFolders should correctly materialize files and folders 1`] = `
Map {
  "/tmp/dendron/bar/alpha/index.md" => "---
id: bar-alpha-id
---
bar.alpha
",
  "/tmp/dendron/bar/alpha/one.md" => "---
id: bar-alpha-one-id
---
bar.alpha.one
",
  "/tmp/dendron/bar/index.md" => "---
id: bar-id
---
bar
",
  "/tmp/dendron/bar/two.md" => "---
id: bar-two-id
---
bar.two
",
  "/tmp/dendron/foo/two/three.md" => "---
id: foo-two-three-id
---
foo.two.three
",
}
`;

exports[`materializeFnames2FilesAndFolders transformTags should replace dots in tags with slashes 1`] = `
Map {
  "/tmp/dendron/tags/body-only.md" => "---
tags: []
---
#body/one
#body/one/two
",
  "/tmp/dendron/tags/fm-and-body.md" => "---
tags:
  - fm/one
---
#body/one
",
  "/tmp/dendron/tags/fm-only.md" => "---
tags:
  - fm/one
---

",
}
`;
